core/lib/contract_verifier/src/lib.rs

```diff
@@ -353,6 +363,98 @@ impl ContractVerifier {
         Ok((info, compiled_identifier))
     }
 
+    /// Returns compilation artifacts for the given request.
+    /// Tries to use the compiler versions from the metadata if they are different from the ones in the request.
+    /// If the metadata is not present, or the versions are the same, uses the ones from the request.
+    async fn get_compilation_artifacts(
+        &self,
+        request: &mut VerificationRequest,
+        detected_metadata: Option<&DetectedMetadata>,
+        bytecode_marker: &BytecodeMarker,
+    ) -> Result<CompilationArtifacts, ContractVerifierError> {
+        match detected_metadata {
+            Some(DetectedMetadata::Cbor {
+                metadata: cbor_metadata,
+                ..
+            }) if !request.req.compiler_versions_match(cbor_metadata) => {
+                let updated_request = request
+                    .req
+                    .clone()
+                    .with_updated_compiler_versions(cbor_metadata);
+
+                match self
+                    .compile(updated_request.clone(), *bytecode_marker)
+                    .await
+                {
+                    Ok(artifacts) => {
+                        tracing::warn!(
+                            request_id = request.id,
+                            request_compiler = request.req.compiler_versions.compiler_version(),
+                            metadata_compiler =
+                                updated_request.compiler_versions.compiler_version(),
+                            request_zk_compiler =
+                                request.req.compiler_versions.zk_compiler_version(),
+                            metadata_zk_compiler =
+                                updated_request.compiler_versions.zk_compiler_version(),
+                            "Updating request compiler versions in the DB."
+                        );
+
+                        // Since we know the version that should be used for the contract, we override one provided in
+                        // the request, so that if this request is interpreted by anyone else, they can get correct
+                        // compilation results
+                        self.update_request_compiler_versions(
```

Comment:
We should still check that the verification passes, right? E.g. it can compile but still not result in a match.